pgmpy/factors/FactorSet.py

```diff
@@ -0,0 +1,191 @@
+        """
+        return factorset_divide(self, factorset)
+
+    def marginalize(self, variables, inplace=True):
+        """
+        Marginalizes the factors present in the factor sets with respect to the given variables.
+
+        Parameters
+        ----------
+        variables: string, list-type
+            name of the variable (or variables) to be marginalized
+
+        inplace: boolean
+            If inplace=True it will modify the factor set itself, else it would create a new factor set
+
+        Examples
+        --------
+        >>> from pgmpy.factors import FactorSet
+        >>> from pgmpy.factors import Factor
+        >>> phi1 = Factor(['x1', 'x2', 'x3'], [2, 3, 2], range(12))
+        >>> phi2 = Factor(['x3', 'x4', 'x1'], [2, 2, 2], range(8))
+        >>> factor_set1 = FactorSet(phi1, phi2)
+        >>> factor_set1.marginalize('x1')
+        """
+        if not isinstance(variables, (list, set, tuple)):
+            variables = [variables]
+
+        factors_to_be_marginalized = set(filter(lambda x: len(set(x.scope()).intersection(variables)) != 0,
+                                                self.factors_set))
+        if not inplace:
+            new_factors_set = self.factors_set - factors_to_be_marginalized
+
+        for factor in factors_to_be_marginalized:
+            variables_to_be_marginalized = list(set(factor.scope()).intersection(variables))
+            if inplace:
+                factor.marginalize(variables_to_be_marginalized, inplace=True)
+            else:
+                new_factors_set.add(factor.marginalize(variables_to_be_marginalized, inplace=False))
+
+        if not inplace:
+            return FactorSet(*new_factors_set)
+
+
+def factorset_product(*factorsets_list):
+    r"""
+    Base method used for product of factor sets.
+
+    Suppose :math:`\vec\phi_1` and :math:`\vec\phi_2` are two factor sets then their product is a another factors set
+    :math:`\vec\phi_3 = \vec\phi_1 \cup \vec\phi_2`.
+
+    Parameters
+    ----------
+    factorsets_list: FactorSet1, FactorSet2, ..., FactorSetn
+        All the factor sets to be multiplied
+
+    Examples
+    --------
+    >>> from pgmpy.factors import FactorSet
+    >>> from pgmpy.factors import Factor
+    >>> from pgmpy.factors import factorset_product
+    >>> phi1 = Factor(['x1', 'x2', 'x3'], [2, 3, 2], range(12))
+    >>> phi2 = Factor(['x3', 'x4', 'x1'], [2, 2, 2], range(8))
+    >>> factor_set1 = FactorSet(phi1, phi2)
+    >>> phi3 = Factor(['x5', 'x6', 'x7'], [2, 2, 2], range(8))
+    >>> phi4 = Factor(['x5', 'x7', 'x8'], [2, 2, 2], range(8))
+    >>> factor_set2 = FactorSet(phi3, phi4)
+    >>> factor_set3 = factorset_product(factor_set1, factor_set2)
+    """
+    if not all(isinstance(factorset, FactorSet) for factorset in factorsets_list):
+        raise TypeError("Input parameters must be factor sets")
+    return reduce(lambda x, y: FactorSet(*(x.factors_set.union(y.factors_set))), factorsets_list)
```

Comment:
@abinashpanda This doesn't look very good. We are creating FactorSet objects in each iteration. What do you think of something like this:
```python3
FactorSet(*set.union(*map(lambda t: t.factors_set, factorsets_list)))
```
Also this is faster. For `factorsets_list ` having 200 elements:
```python3
%timeit FactorSet(*set.union(*map(lambda t: t.factors_set, factorsets_list)))
1000 loops, best of 3: 305 µs per loop

%timeit reduce(lambda x, y: FactorSet(*(x.factors_set.union(y.factors_set))), factorsets_list)
1 loops, best of 3: 178 ms per loop
```